src/libasr/codegen/llvm_utils.cpp

```diff
@@ -589,8 +589,37 @@ namespace LFortran {
         LLVM::CreateStore(*builder, dest_key_mask, dest_key_mask_ptr);
     }
 
-    void LLVMList::check_index_within_bounds(llvm::Value* /*list*/, llvm::Value* /*pos*/) {
+    void LLVMList::check_index_within_bounds(llvm::Value* list, llvm::Value* pos) {
+        llvm::Value* end_point = LLVM::CreateLoad(*builder,
+                                    get_pointer_to_current_end_point(list));
+        llvm::Value* zero = llvm::ConstantInt::get(llvm::Type::getInt32Ty(context),
+                                                   llvm::APInt(32, 0));
 
+        llvm::Function *fn = builder->GetInsertBlock()->getParent();
+        llvm::BasicBlock *thenBB = llvm::BasicBlock::Create(context, "then", fn);
+        llvm::BasicBlock *elseBB = llvm::BasicBlock::Create(context, "else");
+        llvm::BasicBlock *mergeBB = llvm::BasicBlock::Create(context, "ifcont");
+
+        llvm::Value* cond = builder->CreateOr(
+                                builder->CreateICmpSGT(pos, end_point),
+                                builder->CreateICmpSGT(zero, pos));
+        builder->CreateCondBr(cond, thenBB, elseBB);
+        builder->SetInsertPoint(thenBB);
+        {
+            std::string message = "list index out of range";
+            llvm::Value *fmt_ptr = builder->CreateGlobalStringPtr("IndexError: %s\n");
```

Comment:
I think its time to work out a method to provide runtime information (like in this the index and the bounds of the list). Its an easy-medium thing. Would you like to do that?

Reply:
May be do it in a follow up PR.

Reply:
Another pain point is to test whether this error is raised or not for out of range indices. In  CPython we use `pytest.raises` (or something like that, don't remember the exact details). I am not sure what to use here?

Reply:
Yes, we need that, so Thirumalai and I created an issue for this: https://github.com/lcompilers/lpython/issues/1128